tower/src/util/boxed_clone_sync.rs

```diff
@@ -0,0 +1,101 @@
+use super::ServiceExt;
+use futures_util::future::BoxFuture;
+use std::{
+    fmt,
+    task::{Context, Poll},
+};
+use tower_layer::{layer_fn, LayerFn};
+use tower_service::Service;
+
+/// A [`Clone`] + [`Send`] + [`Sync`] boxed [`Service`].
+///
+/// [`BoxCloneSyncService`] turns a service into a trait object, allowing the
+/// response future type to be dynamic, and allowing the service to be cloned and shared.
+///
+/// This is similar to [`BoxCloneService`](super::BoxCloneService) except the resulting
+/// service implements [`Sync`].
+/// ```
+pub struct BoxCloneSyncService<T, U, E>(
+    Box<
+        dyn CloneService<T, Response = U, Error = E, Future = BoxFuture<'static, Result<U, E>>>
+            + Send
+            + Sync,
+    >,
+);
+
+impl<T, U, E> BoxCloneSyncService<T, U, E> {
+    /// Create a new `BoxCloneService`.
+    pub fn new<S>(inner: S) -> Self
+    where
+        S: Service<T, Response = U, Error = E> + Clone + Send + Sync + 'static,
+        S::Future: Send + 'static,
+    {
+        let inner = inner.map_future(|f| Box::pin(f) as _);
+        BoxCloneSyncService(Box::new(inner))
+    }
+
+    /// Returns a [`Layer`] for wrapping a [`Service`] in a [`BoxCloneSyncService`]
+    /// middleware.
+    ///
+    /// [`Layer`]: crate::Layer
+    pub fn layer<S>() -> LayerFn<fn(S) -> Self>
```

Comment:
Part of me is a little skeptical of naming this util type here. It would mean we couldn't change to a dedicated layer type without it being "breaking". Changing this to `impl Layer` could work, but people eventually complain.

Is this what you were referring to @jlizen?

Reply:
I'm not sure I understand the concern. This maps to how `BoxCloneService` and `BoxService` implement it as well. The service has function that returns an unnamed layer function, or else there is a separate layer module. 

Both started out without the dedicated layer and added it in, so perhaps the unnamed function is more vestigial.

Anyway, I think it's probably fine to have both, and libraries can use the named one if they want to expose it to the caller. Probably more confusing to have the API diverge from the other type erasure wrappers.

Fwiw, for my purposes in that reqwest commit, I only need the service wrapper anyway. I've switched over to composing with generics, and then if custom layers are provided, wrapping the final thing in a `BoxCloneSyncService` to avoid polluting the built client with long generic bounds everywhere.

Reply:
Ah, I didn't realize other types have two separate layer constructors. Hm, that seems like an eventual place to improve, but oh well.